# Grade Prediction using Machine Learning

## Project Objective

The objective of this project is to build a machine learning regression model that predicts a student's final grade based on academic and performance-related features.

The trained model will later be saved and integrated into a Flask web application to generate predictions dynamically.

In [1]:
# Ignore unnecessary warnings
import warnings
warnings.filterwarnings("ignore")

# Data manipulation
import pandas as pd
import numpy as np

# Data visualization
import matplotlib.pyplot as plt
import seaborn as sns

## 1. Load the Dataset

The student performance dataset was loaded into a Pandas DataFrame. This dataset contains academic, personal, and environmental factors that may influence a student's final exam score.

In [2]:
df = pd.read_csv("../dataset/StudentPerformanceFactors.csv")

df.head()

,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23,84,Low,High,No,7,73,Low,Yes,0,Low,Medium,Public,Positive,3,No,High School,Near,Male,67
1,19,64,Low,Medium,No,8,59,Low,Yes,2,Medium,Medium,Public,Negative,4,No,College,Moderate,Female,61
2,24,98,Medium,Medium,Yes,7,91,Medium,Yes,2,Medium,Medium,Public,Neutral,4,No,Postgraduate,Near,Male,74
3,29,89,Low,Medium,Yes,8,98,Medium,Yes,1,Medium,Medium,Public,Negative,4,No,High School,Moderate,Male,71
4,19,92,Medium,Medium,Yes,6,65,Medium,Yes,3,Medium,High,Public,Neutral,4,No,College,Near,Female,70


In [3]:
df.shape

(6607, 20)

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6607 entries, 0 to 6606
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype
---  ------                      --------------  -----
 0   Hours_Studied               6607 non-null   int64
 1   Attendance                  6607 non-null   int64
 2   Parental_Involvement        6607 non-null   str  
 3   Access_to_Resources         6607 non-null   str  
 4   Extracurricular_Activities  6607 non-null   str  
 5   Sleep_Hours                 6607 non-null   int64
 6   Previous_Scores             6607 non-null   int64
 7   Motivation_Level            6607 non-null   str  
 8   Internet_Access             6607 non-null   str  
 9   Tutoring_Sessions           6607 non-null   int64
 10  Family_Income               6607 non-null   str  
 11  Teacher_Quality             6529 non-null   str  
 12  School_Type                 6607 non-null   str  
 13  Peer_Influence              6607 non-null   str  
 14  Physical_Activity  

### Observation

The dataset contains **6,607 records** and **20 features**, including both numerical and categorical variables.

Most features have complete data; however, **Teacher_Quality**, **Parental_Education_Level**, and **Distance_from_Home** contain missing values. These missing values must be handled during data preprocessing before training the machine learning model.

The target variable, **Exam_Score**, is numerical, making this dataset suitable for a regression problem.

## 4. Missing Values Analysis

The dataset was checked for missing values to identify incomplete records before data preprocessing. Missing values must be handled because machine learning algorithms cannot directly work with incomplete data.

In [5]:
df.isnull().sum()

Hours_Studied                  0
Attendance                     0
Parental_Involvement           0
Access_to_Resources            0
Extracurricular_Activities     0
Sleep_Hours                    0
Previous_Scores                0
Motivation_Level               0
Internet_Access                0
Tutoring_Sessions              0
Family_Income                  0
Teacher_Quality               78
School_Type                    0
Peer_Influence                 0
Physical_Activity              0
Learning_Disabilities          0
Parental_Education_Level      90
Distance_from_Home            67
Gender                         0
Exam_Score                     0
dtype: int64

## 5. Handle Missing Values

The missing values in the categorical features were replaced using **Mode Imputation**. Since these features contain text-based categories, the most frequently occurring value (mode) was used to fill the missing entries while preserving the dataset.

In [8]:
df["Teacher_Quality"] = df["Teacher_Quality"].fillna(
    df["Teacher_Quality"].mode()[0]
)

df["Parental_Education_Level"] = df["Parental_Education_Level"].fillna(
    df["Parental_Education_Level"].mode()[0]
)

df["Distance_from_Home"] = df["Distance_from_Home"].fillna(
    df["Distance_from_Home"].mode()[0]
)

df.isnull().sum()


Hours_Studied                 0
Attendance                    0
Parental_Involvement          0
Access_to_Resources           0
Extracurricular_Activities    0
Sleep_Hours                   0
Previous_Scores               0
Motivation_Level              0
Internet_Access               0
Tutoring_Sessions             0
Family_Income                 0
Teacher_Quality               0
School_Type                   0
Peer_Influence                0
Physical_Activity             0
Learning_Disabilities         0
Parental_Education_Level      0
Distance_from_Home            0
Gender                        0
Exam_Score                    0
dtype: int64

### Observation

The missing values were successfully handled using **Mode Imputation**. After preprocessing, all features contain complete data with no missing values, making the dataset suitable for further analysis and machine learning model training.

## 6. Duplicate Values Analysis

The dataset was checked for duplicate records to ensure data quality. Duplicate entries can introduce bias during model training and negatively affect the performance of machine learning models.

In [9]:
df.duplicated().sum()

np.int64(0)

### Observation

No duplicate records were found in the dataset. This confirms that each row represents a unique student record, and no duplicate entries need to be removed before model training.